content/blog/2020-12-20-loop-flatten/index.md

```diff
@@ -0,0 +1,199 @@
+```javascript
+for (int i = 0; i < n; i++) { // unflattened “perfect” nested loop
+    // no statement here
+    for (int j = 0; j < m; j++) {
+      Stmt(f(i, j))
+```
+
+My naïve pass is available in [this link](https://github.com/hj424/bril/tree/master/tasks/project/llvm-pass-skeleton).
+
+## Evaluation
+1. Correctness <br/>
+As for the correctness of the pass, we should focus on two points: the functionality of the optimized code and the correctness of the unrolled loop. These two points are being checked by running the simple benchmarks with the clang profiling tools. For more details, please check out [this link](https://clang.llvm.org/docs/SourceBasedCodeCoverage.html).
+
+ As documented in the readme file of the code repository, you can run the code with my bash script.
+
+To elaborate the evaluation, let me use the following source code as an example: 
+
+```javascript
+#include <stdio.h>
+#include <stdlib.h>
+
+#define N 11
+#define M 20
+
+int main() {
+  int init_val = 30;
+  int res[M*N];
+  LOOP:
+  for (int i = 0; i < N; i++) {
+    for (int j = 0; j < M; j++) {
+      res[i*M+j] = init_val;
+    }
+  }
+  printf("Multiplication res: %d; \n", res[0]);
+  return 0;
+}
+```
+
+After optimizing this loop with my pass, the runtime profiling is showed here.
+
+```javascript
+Multiplication res: 30; 
+    1|       |#include <stdio.h>
+    2|       |#include <stdlib.h>
+    3|       |
+    4|      2|#define N 11
+    5|    441|#define M 20
+    6|       |
+    7|      1|int main() {
+    8|      1|  int init_val = 30;
+    9|      1|  int res[M*N];
+   10|      1|  LOOP:
+   11|      2|  for (int i = 0; i < N; i++) {
+   12|    221|    for (int j = 0; j < M; j++) {
+   13|    220|      res[i*M+j] = init_val;
+   14|    220|    }
+   15|      1|  }
+   16|      1|  printf("Multiplication res: %d; \n", res[0]);
+   17|      1|  return 0;
+   18|      1|}
+```
+The first line is the sample result I printed out to verify the correctness of the optimized code. And then the following code is the optimized one with the updated inner loop bound. As you can see from line 11, the outer loop was being executed only 1 time. The number is 2 here means the second time, it checks the loop bound and exits the outer loop. And accordingly, the inner loop is being executed `N*M` times, which is 220 from line 13. This naïve pass is being tested with other small benchmarks I created with perfect nested loop and passed all the tests.
+
+2. Performance gain<br/>
+As we mentioned before, the loop flattening pass itself does not offer any benefits for performance. But this technique helps to better explore the parallelism when combining with the loop unrolling technique. Specifically, we can apply larger unroll parameters to further explore the parallelism across the loop iterations. Also, this technique is used to support other optimizing pragmas (like loop pipelining in HLS tools) for better performance.

```

Comment:
It would have been interesting, FWIW, to have measured the impact on these downstream optimizations.